docs/developer-docs/security/rust-canister-development-security-best-practices.md

```diff
@@ -262,8 +262,151 @@ Every inter-canister call is guaranteed to receive a response, either from the c
+
+The naïve recommendation to address the security concern described above would be to avoid traps. However, that can be very difficult to achieve due to the following reasons: 
+
+- The implementation can be involved and could panic due to bugs, such as index out-of-bounds errors or panics (expect, unwrap) that should supposedly never happen.
+
+- It is hard to make sure the callback or cleanup doesn’t run into the instruction limit (and thus trap), because the number of instructions required can in general not be predicted and may e.g. depend on the data being processed. 
+
+Due to these reasons, while it is easy to recommend “avoiding traps”, this is actually hard to achieve in practice. So in our view, code should be written so that it can deal even with unexpected traps due to bugs or hitting the instruction limits. We discuss two approaches:
+
+1. Do simple cleanups 
+1. Do “journaling”
+
+In the first approach, the cleanup callback is used to recover from unexpected panics. This can work, but has several drawbacks: 
+- The cleanup itself could panic, in which case one is in the initial problematic situation again. The risk may be acceptable for simple cleanups, but as discussed above it is hard to write code that never panics, especially if it is somewhat complex. 
+- Currently, Motoko does not allow setting custom cleanup callbacks. Cleanup is used internally by Motoko to do some cleanups.
+- As discussed above, the Rust CDK has a feature that automatically releases local variables in cleanup, which [can be used to release locks](https://internetcomputer.org/docs/current/developer-docs/security/rust-canister-development-security-best-practices#be-aware-that-there-is-no-reliable-message-ordering). Since only one cleanup callback can be defined, any custom cleanup would currently have to implement that feature itself if needed, making this currently hard to use and understand.
+
+We proceed to discuss “journaling”, which is currently our recommended way of addressing the problem at hand. 
+
+### Journaling
+
+Journaling can be used for ensuring that tasks are completed correctly in an asynchronous context, where any instruction or async task can fail. Journaling is generally useful in any security critical application canister on the IC. The journaling concept we describe here is inspired and adapted from journaling in file systems.
+
+A journal is a chronological list of the records and is kept in the canister’s storage. It keeps track of tasks before they begin and when they are completed. Before each failable task, the journal records the intent to execute the task, and after the task, the journal records the result. Creating a record in the journal is called “journaling”.  For example, to make an unreliable async call to a ledger:
```

Comment:
How is a journal replay triggered? Is it by a user message retrying the procedure? If that's the case, should the user message also be logged in the journal, so that the successfully executed parts of handling the call are not done twice? If not (e.g., if it's a heartbeat that triggers the recovery), how does the user learn about the outcome of its call?

Reply:
If the call should be idempotent, the journal should contain enough information so that if the user retries, the app can determine which subtasks were successful and do not need to be redone. The app should then attempt to recover and continue the task. 

A recovery process could also be triggered on heartbeat. In this case, I would consider two options:
* The recovery process only resolves the status of unresolved entries and takes no actions. It is up to the user to decide if they want to continue the task or abort/clean up.
* The recovery process not only resolves unresolved entries, but it also attempts to continue the task. In this case the app would need to allow the user to check the status of the task. If the task is idempotent, the user could resubmit and see the status in the response.

Reply:
I've added a description of retries.